src/modm/driver/encoder/encoder_impl.hpp

```diff
@@ -0,0 +1,49 @@
+#ifndef MODM_ENCODER_HPP
+#error "Don't include this file directly, use 'encoder.hpp' instead!"
+#endif
+
+template<typename PH_A, typename PH_B, modm::encoder::Division_t DIVISION, typename UnderlyingType>
+inline void
+modm::Encoder<PH_A, PH_B, DIVISION, UnderlyingType>::update()
+{
+	static_assert(std::is_signed<UnderlyingType>::value and std::is_integral<UnderlyingType>::value,
+				  "UnderlyingType must be signed integral");
+	// Tare on power-on
+	UnderlyingType raw = 0;
+	UnderlyingType diff;
+
+	if (PH_A::read()) raw = 3;
+	if (PH_B::read()) raw ^= 1;  // convert gray to binary
+
+	diff = raw_last - raw;
+	if (diff & 1)
+	{
+		raw_last = raw;               // store new as next last
+		enc_delta += (diff & 2) - 1;  // bit 1 = direction (+/-)
+	}
+}
+
+template<typename PH_A, typename PH_B, modm::encoder::Division_t DIVISION, typename UnderlyingType>
+UnderlyingType
+modm::Encoder<PH_A, PH_B, DIVISION, UnderlyingType>::getValue()
+{
+	static_assert(std::is_signed<UnderlyingType>::value and std::is_integral<UnderlyingType>::value,
+				  "UnderlyingType must be signed integral");
+	cli();
+	UnderlyingType val = enc_delta;  // counts since last call
+	switch (DIVISION)
+	{
+		case encoder::Division::ONE:
+			enc_delta = 0;
+			break;
+		case encoder::Division::TWO:
+			enc_delta = val & 1;
+			val >>= 1;
+			break;
+		case encoder::Division::FOUR:
+			enc_delta = val & 3;
+			val >>= 2;
+	}
```

Comment:
> getValue() could have been specialized for each value in < DIVISION >. Because of the difference of just 3 tiny lines, i found the it better using a switch-statement.

This looks like it could be generalized by passing an integer instead of an enum?

Reply:
> > getValue() could have been specialized for each value in < DIVISION >. Because of the difference of just 3 tiny lines, i found the it better using a switch-statement.
> 
> This looks like it could be generalized by passing an integer instead of an enum?

In my opinion that's the last thing left. Some for tomorrow or someone else.